src/mpi/comm/commutil.c

```diff
@@ -1184,6 +1206,10 @@ int MPIR_Comm_delete_internal(MPIR_Comm * comm_ptr)
     }
 
   fn_exit:
+    if (has_unmatched_message) {
+        MPIR_ERR_SET(mpi_errno, MPI_ERR_OTHER, "**commhasunmatched");
```

Comment:
Maybe enhanced it by telling the user how many unmatched messages?

Reply:
Can do. Also noticed in experiments that this check won't apply to builtin comms since they don't get "deleted". Might need to convert to helper so it can get used for those comms, as well.

Reply:
I take that back. Those comms will get deleted. Need to look closer at why the error didn't show up with my simple test.

Reply:
Builtin comms are freeed in MPI_Finalize, it may not follow the usual error failure path.

Reply:
Yup. Added a patch to capture errors from the builtin finalize function. Now my experiment works with builtin and derived comms. I should adapt it to the testsuite.